engine/consensus/approvals/assignment_collector_statemachine.go

```diff
@@ -0,0 +1,264 @@
+package approvals
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/engine/consensus"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+var (
+	ErrInvalidCollectorStateTransition = errors.New("invalid state transition")
+	ErrDifferentCollectorState         = errors.New("different state")
+)
+
+type AssignmentCollectorStateMachine struct {
+	AssignmentCollectorBase
+
+	sync.Mutex
+	collector atomic.Value
+}
+
+func (asm *AssignmentCollectorStateMachine) atomicLoadCollector() AssignmentCollectorState {
+	return asm.collector.Load().(*atomicValueWrapper).collector
+}
+
+// atomic.Value doesn't allow storing interfaces as atomic values,
+// it requires that stored type is always the same so we need a wrapper that will mitigate this restriction
+// https://github.com/golang/go/issues/22550
+type atomicValueWrapper struct {
+	collector AssignmentCollectorState
+}
+
+func NewAssignmentCollectorStateMachine(collectorBase AssignmentCollectorBase) *AssignmentCollectorStateMachine {
+	sm := &AssignmentCollectorStateMachine{
+		AssignmentCollectorBase: collectorBase,
+	}
+
+	// by default start with caching collector
+	sm.collector.Store(&atomicValueWrapper{
+		collector: NewCachingAssignmentCollector(collectorBase),
+	})
+	return sm
+}
+
+// ProcessIncorporatedResult starts tracking the approval for IncorporatedResult.
+// Method is idempotent.
+// Error Returns:
+//  * no errors expected during normal operation;
+//    errors might be symptoms of bugs or internal state corruption (fatal)
+func (asm *AssignmentCollectorStateMachine) ProcessIncorporatedResult(incorporatedResult *flow.IncorporatedResult) error {
+	for {
+		collector := asm.atomicLoadCollector()
+		currentState := collector.ProcessingStatus()
+		err := collector.ProcessIncorporatedResult(incorporatedResult)
+		if err != nil {
+			return err
+		}
+		if currentState != asm.ProcessingStatus() {
```

Comment:
```suggestion
		// since the results and approvals are processed concurrently, if the currentState has
		// changed, it means we let the wrong collect process the result. In this case, we just
		// re-process the result again, until there was no state transition during the process,
		// which ensured the result has been processed by the correct collector.
		if currentState != asm.ProcessingStatus() {
```